src/docfx/lib/ProcessUtility.cs

```diff
@@ -251,5 +415,47 @@ private static async Task<T> RetryUntilSucceed<T>(string name, Func<Exception, b
                 }
             }
         }
+
+        private static string GetLockFilePath(string lockName)
+        {
+            Directory.CreateDirectory(AppData.MutexRoot);
+            var lockPath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(lockName) + "-rw");
+            return lockPath;
+        }
+
+        private static LockInfo FilterExpiredAcquirers(LockInfo lockInfo)
+        {
+            Debug.Assert(lockInfo != null);
+
+            lockInfo.AcquiredBy.RemoveAll(r => DateTime.UtcNow - r.Date > TimeSpan.FromSeconds(_defaultLockExpireTimeInSecond));
+
+            if (!lockInfo.AcquiredBy.Any())
+            {
+                lockInfo.Type = LockType.None;
+            }
+
+            return lockInfo;
+        }
+
+        private enum LockType
+        {
+            None,
+            Shared,
+            Exclusive,
+        }
+
+        private class LockInfo
+        {
+            public LockType Type { get; set; }
+
+            public List<Acquirer> AcquiredBy { get; set; } = new List<Acquirer>();
```

Comment:
how about `readers` because acquired could both mean acquired for read and acquired for write.

Reply:
i would like to keep the acquirer  info even it's write lock acquiring, debugging would be easy